### PR TITLE
83 Links 2.0

### PR DESCRIPTION
**Pull request by eiabea**
AC:
- [x]  links can point to jira projects or epics
- [x]  editable links
- [x]  publish should show where it would be published
- [x]  info of project should display link
- [x]  help should have a description with epics
- [x]  bump up coverage

### CODE REVIEW

**Comment by eiabea** [comment]
I encountered a problem with the initial idea:
- Adding a new key to the `IJiraLink` interface like `epic` would result in a config version upgrade, therefore a new logic for upgrading the config file

I checked jira on how an epic is handled and it turns out that an epic is like a story (syntax-wise) and is always bundled with a project. So i came up with the idea of enriching the current `key` parameter of `IJiraLink` interface with a custom format `key: PROJECT/EPIC-1` and send this to the endpoint and let it be handled there

@matthiaseigner @lombad please check out the changes so far and share your opinions :baby:

**Comment by lombad** [comment]
> I encountered a problem with the initial idea:
> 
>     * Adding a new key to the `IJiraLink` interface like `epic` would result in a config version upgrade, therefore a new logic for upgrading the config file would be necessary
> 
> 
> I checked jira on how an epic is handled and it turns out that an epic is like a story (syntax-wise) and is always bundled with a project. So i came up with the idea of enriching the current `key` parameter of `IJiraLink` interface with a custom format `key: PROJECT/EPIC-1` and send this to the endpoint and let it be handled there
> 
> @matthiaseigner @lombad please check out the changes so far and share your opinions baby

If that works, which we should check the JIRA side first. But it sounds like a legit solution without the need of separating the logic on the cli side. Which will anyway not be needed if we think of Epic as the same as a Project later on and as discussed lately.

**Comment by matthiaseigner** [comment]
> > I encountered a problem with the initial idea:
> > ```
> > * Adding a new key to the `IJiraLink` interface like `epic` would result in a config version upgrade, therefore a new logic for upgrading the config file would be necessary
> > ```
> > 
> > 
> > I checked jira on how an epic is handled and it turns out that an epic is like a story (syntax-wise) and is always bundled with a project. So i came up with the idea of enriching the current `key` parameter of `IJiraLink` interface with a custom format `key: PROJECT/EPIC-1` and send this to the endpoint and let it be handled there
> > @matthiaseigner @lombad please check out the changes so far and share your opinions baby
> 
> If that works, which we should check the JIRA side first. But it sounds like a legit solution without the need of separating the logic on the cli side. Which will anyway not be needed if we think of Epic as the same as a Project later on and as discussed lately.

I would propose that we use the name "issue" instead of "epic". If jira handles epics as issue with the type epic we can support adding hours to issues as well (probably not needed but a free feature anyways)

as discussed additionally to that I would add a new user story to jira about handling config migrations

**Comment by eiabea** [comment]
> > > I encountered a problem with the initial idea:
> > > ```
> > > * Adding a new key to the `IJiraLink` interface like `epic` would result in a config version upgrade, therefore a new logic for upgrading the config file would be necessary
> > > ```
> > > 
> > > 
> > > I checked jira on how an epic is handled and it turns out that an epic is like a story (syntax-wise) and is always bundled with a project. So i came up with the idea of enriching the current `key` parameter of `IJiraLink` interface with a custom format `key: PROJECT/EPIC-1` and send this to the endpoint and let it be handled there
> > > @matthiaseigner @lombad please check out the changes so far and share your opinions baby
> > 
> > 
> > If that works, which we should check the JIRA side first. But it sounds like a legit solution without the need of separating the logic on the cli side. Which will anyway not be needed if we think of Epic as the same as a Project later on and as discussed lately.
> 
> I would propose that we use the name "issue" instead of "epic". If jira handles epics as issue with the type epic we can support adding hours to issues as well (probably not needed but a free feature anyways)
> 
> also I think the issue key should not be encoded in an arbitrary way (separated by /) but should be sent in an own parameter
> 
> as discussed additionally to that I would add a new user story to jira about handling config migrations

Thank you for the feedback!

> as discussed additionally to that I would add a new user story to jira about handling config migrations

I will do that!

> also I think the issue key should not be encoded in an arbitrary way (separated by /) but should be sent in an own parameter

I will leave the `key` parameter as it is and and an additional `issue` parameter, which will not require a upgrade logic, at least not in the newer-direction

New `IJiraLink` interface:

```
export interface IJiraLink extends IIntegrationLink {
  username: string;
  hash: string;
  endpoint: string;
  key: string;
  issue: string;
}
```

File: helper/question.ts [hunk 1]
```diff
@@ -130,47 +131,65 @@ export class QuestionHelper {
...
+        .toString("base64");
+    }
 
     const projectName: string = project.name;
 
     const link: IJiraLink = {
-      // Tailing slash is ensured by the filter
-      endpoint: `${host}rest/gittt/latest/`,
+      host,
+      // TODO use constant for endpoint version
```
Comment (matthiaseigner) [inline helper/question.ts]: should we leave this todo open for now?

File: test/unit/app.test.ts [hunk 2]
```diff
@@ -1965,940 +1968,1511 @@ describe("App", function () {
...
-        await mockedApp.linkAction(new Command());
+          mockedHelper.FileHelper = class {
+            public static getHomeDir = sinon.stub().returns("/home/test");
+            public configDirExists = sinon.stub().resolves(true);
+            public isConfigFileValid = sinon.stub().resolves(true);
+            public addOrUpdateLink = addOrUpdateLinkStub;
+            public findLinkByProject = sinon.stub().resolves(
+              {
+                endpoint: "/rest/gittt/latest/",
+                hash: "bW9ja2VkOm1vY2tlZA==",
```
Comment (matthiaseigner) [inline test/unit/app.test.ts]: real password?

Comment (eiabea) [reply test/unit/app.test.ts]: Yes, it is the hash of `mocked:mocked`

File: helper/question.ts (hunk 1 above)
Comment (eiabea) [reply helper/question.ts]: Fixed in [here](https://github.com/codecovevienna/gittt-cli/pull/49/commits/d0813cd06c52b294a4bbdda842bb94041bbb02f1)

**Pull request merged**